### PR TITLE
Feature/9 prefetching records

### PR DESCRIPTION
**Pull request by wem**

### CODE REVIEW

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt [hunk 1]
```diff
@@ -110,20 +125,18 @@ abstract class AbstractKinesisConsumerVerticle : CoroutineVerticle() {
         }
     }
 
-    private fun startPolling(fetchStartPosition: QueryPosition) {
+    private suspend fun startPolling(fetchStartPosition: QueryPosition) {
         logger.debug { "Start polling for $consumerInfo" }
         var positionToFetch = fetchStartPosition
+
+        // We not wait on first fetch
+        recordFetcher.fetchNextRecords(positionToFetch, false)
```
Comment (snakeru) [inline src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt]: Why no waiting? Does it brinch much?

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt [hunk 2]
```diff
@@ -110,20 +125,18 @@ abstract class AbstractKinesisConsumerVerticle : CoroutineVerticle() {
...
             while (isActive && running) {
                 runCatching {
-                    recordFetcher.fetchNextRecords(positionToFetch)
+                    recordFetcher.getNextRecords()
                 }.onSuccess { recordsResponse ->
-                    // We throttle as coroutine, so when the record processing will take longer than the configured
-                    // interval, we can continue polling immediately afterwards.
-                    val throttleJob = launch {
-                        delay(options.kinesisPollIntervalMillis)
-                    }
```
Comment (snakeru) [inline src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt:125]: Unrelated comment: Michel, have you looked at my feedback for the previous (large) PR?

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt [hunk 3]
```diff
@@ -134,13 +147,24 @@ abstract class AbstractKinesisConsumerVerticle : CoroutineVerticle() {
...
                     }.getOrElse { throwable ->
-                        handleConsumerFailure(throwable, nextPositionIfDeliveredSuccessful, positionToFetch)
+                        handleConsumerFailure(throwable, nextPositionIfDeliveredSuccessful, positionToFetch).also { positionAfterFailure ->
+                            positionAfterFailure?.let {
+                                // Only restart fetching if it's really needed
+                                if (positionAfterFailure != nextPositionIfDeliveredSuccessful) {
+                                    recordFetcher.restartFetching(it)
+                                }
+                            }
+                        }
```
Comment (snakeru) [inline src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt]: This whole construct is somewhat hard to understand. We manipulate three positions (positionToFetch, nextPositionIfDeliveredSuccessful, positionAfterFailure) to calculate the fourth it seems and also start/stop fetching in the process. May be splitting this into some (possibly inlined) methods might help. Or may be an illustration to describe how these relate to each other.

--

BTW - why this whole store with re-fetching? What kind of failure scenario is envisioned?

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcher.kt [hunk 4]
```diff
@@ -14,27 +15,50 @@ class RecordFetcher(
...
+            // CancelFetchJobException should be catched by caller. This should never happen on deffered.await / getNextRecords()
+            fetchNextRecords(positionToFetch)
+        }
+    }
+
     /**
-     * Fetching of records. Catches the case when the shard iterator did expire. In this case, the iterator will be
-     * refreshed by the [positionToFetch.sequenceNumber].
+     * Finally fetching of records. Catches the case when the shard iterator did expire. In this case, the iterator will be
+     * refreshed by the [positionToFetch.sequenceNumber] and retried to fetch afterwards.
```
Comment (snakeru) [inline src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcher.kt]:
```suggestion
     * refreshed by the [positionToFetch.sequenceNumber] and fetch will be retried afterwards.
```

File: src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt [hunk 5]
```diff
@@ -74,9 +77,106 @@ internal class RecordFetcherTest : AbstractVertxTest() {
...
+            }
+        }
+
+        val sut = createFetcher(kinesisClientMock)
+
+        sut.fetchNextRecords(queryPosition)
+
+        val previousPosition = QueryPosition(
+            ShardIterator(previousShardIterator),
+            SequenceNumber("sequence-number", SequenceNumberIteratorPosition.AFTER)
```
Comment (snakeru) [inline src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt]: unexpected that AFTER constant is used to create a "previous" value.

File: src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt [hunk 6]
```diff
@@ -74,9 +77,106 @@ internal class RecordFetcherTest : AbstractVertxTest() {
...
+        sut.fetchNextRecords(queryPosition)
+        sut.getNextRecords()
+
+        val end = System.currentTimeMillis()
+        val diff = end - start
+        diff.shouldBeGreaterThanOrEqual(delay)
     }
+
+    @Test
+    internal fun exception_case_restart_fetching(testContext: VertxTestContext) = asyncTest(testContext) {
```
Comment (snakeru) [inline src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt]: even after reading this test I still can't understand the underlying logic. I think more a docsting or possibly refactoring is needed.

File: src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt [hunk 7]
```diff
@@ -74,9 +77,106 @@ internal class RecordFetcherTest : AbstractVertxTest() {
...
+        val response = sut.getNextRecords()
+        response.nextShardIterator().shouldBe(previousShardIterator)
+        verify(kinesisClientMock, times(2)).getRecords(any<Consumer<GetRecordsRequest.Builder>>())
+    }
+
+    @Test
+    internal fun wrong_fetching_access_order(testContext: VertxTestContext) = asyncTest(testContext) {
+        val kinesisClientMock = mock<KinesisAsyncClient> { }
+        val sut = createFetcher(kinesisClientMock)
+        assertFailsWith<RecordFetcher.WrongFetchingOrderException> {
```
Comment (snakeru) [inline src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt]: Perhaps some default value is in effect? It is totally not clear why such sequence of actions should lead to an exception

**Review by snakeru — CHANGES_REQUESTED** [review]

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt (hunk 1 above)
Comment (wem) [reply src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt]: As this is the first fetch, directly after the consumer is deployed. This way, we save time (default 1 second)

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt (hunk 2 above)
Comment (wem) [reply src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt:125]: As discussed via Slack that we merge it, no. I will catch up.

File: src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt (hunk 3 above)
Comment (wem) [reply src/main/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/AbstractKinesisConsumerVerticle.kt]: Good point about the readability. Will refactor it. The orchestra has a configurable failure handling. Either the fetching will continue and ignores any kind of failure, or retries from the failed record (with the correct exception thrown and contains necessary information), or the records get re-fetched from Kinesis according the last used iterator. But the exception handling i plan to refactor. It works but i don't like it. Let us discuss about, ok?

File: src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt (hunk 7 above)
Comment (wem) [reply src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt]: Yes there is an order. To get next records without fetching (on known position) makes no sense. We can also return a nullable result for sure, but this not explicit indicates a mistake. Basically this exception handling and test is especially to cover regressions.

**Comment by wem** [comment]
Any objections should be done now.

**Review by wem — COMMENTED** [review]
Any objections should be done now.

**Pull request merged**